### ISSUE TITLE
Clarify Java API license

### ISSUE THREAD

**Issue by dlmarion** [collectd/collectd#1670]
Would it be possible to publish the api jar to Maven central?

**Comment by rubenk** [comment]
Hi,

I honestly don't know. None of us is familiar with Maven. Next to that the java bindings use JNI and can only be build in place together with collectd.

Could you explain the advantage of doing this?

**Comment by rubenk** [comment]
Closing due to lack of response.

**Comment by dlmarion** [comment]
Sorry, didn't realize you were waiting on a response. The advantage is for anyone writing a java plugin for collectd using Maven as their build environment. The java plugin needs the collectd java api jar to compile successfully and right now the only way to do that is either 1) install collectd on the same machine, or 2) include the collectd api jar file in the java plugin maven module. I could not determine the license for the java api, so I don't know that 2 is feasible.

**Comment by rubenk** [comment]
Thanks for your response.
Is installing Collectd on the machine a problem in practice? You'll need to to that anyway if you want to use the api.

As for the license, it's MIT, see https://github.com/collectd/collectd/blob/master/bindings/java/org/collectd/api/CollectdConfigInterface.java for example.

I'm not opposed to your idea, but someone will have to do the work of building the jar and pushing it to Maven. Is this something the Collectd maintainers have to do or could this be done by someone else too?

**Comment by dlmarion** [comment]
Thanks for the link to the source file. I was confused with the language in the COPYING file. Specifically:
```
collectd consists of a daemon and numerous plugins. The daemon is licensed
under the "MIT License"; its source files are located at src/daemon/. The
plugins are licenses individually, please check the top of the plugin's source
file(s) to see which license applies. The majority of plugins is licensed
either under the "MIT License" or the "GNU General Public License".
```
It does not mention the bindings directory at all. Currently we have to install CollectD to build the plugin. I was thinking of distributing the version of the collectd api jar file that we depend on with our source, but did not because of the license confusion. It sounds as if this is ok to do with the MIT license.

**Comment by rubenk** [comment]
Hmm yes, COPYING could mention the bindings as well.

Would including the Collectd api jar with your source solve your immediate issue? In that case I'll close this, since unless someone steps forward to maintain the jar on Maven Central I'm afraid this is not going to happen. Unfortunately we already have our hands full.

**Comment by dlmarion** [comment]
I think making the license for the java API more clear would resolve this issue. Something to the effect that the Java api source is under the MIT license in its entirety. The MIT license states that I can redistribute it, so I think I am covered there. However, also explicitly stating it in would not hurt. With the mix of licenses in use for the different components, I would have to check each source file in the java api for each version going forward to determine if I could still redistribute the jar.

**Comment by rubenk** [comment]
I agree. Would you like to send a PR for this?

**Comment by dlmarion** [comment]
I should be able to do that later today. Thanks.

**Comment by dlmarion** [comment]
I took some time to look over the source files in the bindings directory. For the java api, most of the source files have the MIT style license, and the last commit for these files say "Relicense to MIT license."  The Datasource, Notification, PluginData, and ValueList classes contain a GPLv2 header. I'm not a lawyer, but I did try to find out what this means. I came across https://en.wikipedia.org/wiki/License_compatibility#GPL_compatibility and based on that it would seem to indicate that the entire java api is now GPL'd.

**Comment by rubenk** [comment]
As far as I can tell everything written by @octo and @tokkee has been relicensed to MIT. Datasource.java and others have been written by Hyperic.

As for what this means for the whole Java api, I'm not a lawyer either, so the standard advise is "please get legal advise before you make any decisions".

Perhaps @octo can tell more about the license of the bindings?

**Comment by dlmarion** [comment]
Well, for anyone that creates a Java plugin using the collectd java api, I think that their plugin has to use a GPL-compatible license as some of the GPL licensed objects are part of the API. That's my interpretation anyway. I don't like the idea of updating the COPYING document and giving advice to people that might be incorrect. I also don't like the idea of doing nothing and leaving others in the same situation. Hmmm....

**Comment by rubenk** [comment]
Yeah you're right. Let's wait for @octo to chime in.

**Comment by dlmarion** [comment]
If nothing else is done, I think either a 'license' label should be added to this issue or change the subject to something like 'Java API license' so that this conversation can be found easily.

**Comment by rubenk** [comment]
Your wish is my command :)

**Comment by ctubbsii** [comment]
Following this conversation, it seems that redistribution (to Maven Central) is possible (under the MIT license), but it's still better if it were to come from the authoritative CollectD community. Perhaps if none of the existing contributors know how to make that happen, they might be willing to accept another contributor into the community in order to make it happen from within that community?

I know personally, I find it annoying when I have to disambiguate between multiple third-party publications of a dependency instead of the third-parties working with the upstream community to do it somewhat authoritatively.

**Comment by rubenk** [comment]
Yes, we're absolutely willing to accept help. Our Java plugin could use some love too.

**Comment by tokkee** [comment]
A few comments from my side (no legal advice, IANAL):

- Strictly speaking, you'll have to check the licenses of all individual files on any new release anyway. The COPYING file can go out of sync and the information from the individual files should be more trusted (heck, in theory, even a single file could have multiple licenses :-/)
- When combining files with different licenses, the following applies:
  - Each single source file (on it's own) can be distributed according to its respective license.
  - A binary built from these sources is subject to the terms of all licenses about combined works. The licenses have to be compatible and, generally, the stricter terms will have to be applied (without invalidating requirements from the other licenses). For MIT and GPL, GPL implies stricter rules.
  - When redistributing source or binaries, I'd highly recommend listing all involved licenses.

As for distributing things through Maven, it should be prevented, if possible, to duplicate the code into other projects. I have no idea how distributing to Maven works but if you're able to build a release process that pulls from the official repo, it would seem like the best option to me. Even better, if that could be hooked into Jenkins. /cc @mfournier for that.

**Comment by ctubbsii** [comment]
Sonatype has guides for publishing to Maven Central for open source projects: http://central.sonatype.org/pages/ossrh-guide.html

**Comment by octo** [comment]
Sorry for the late reply. Essentially, what @tokkee said. 

The vast majority of `.java` files are licensed under the MIT license. Four files, namely

*   `DataSource.java`,
*   `Notification.java`,
*   `PluginData.java` and
*   `ValueList.java`

have been copied from [jcollectd](https://github.com/hyperic/jcollectd) and are licensed under the GPL2. A `.jar` built from these sources should therefore be considered GPL2 licensed.

Those classes were initially written by @dougm who works for VMware. I assume that VMware holds the licensing rights to those files now, since they bought Hyperic / SpringSource. Doug, any chance of relicensing these four files to the MIT license to simplify this situation?

Best regards,
—octo

**Comment by dougm** [comment]
@octo yes, certainly possible, I've been meaning to get the license change approved for the hyperic/jcollectd repo too.  Any interest in moving jcollectd under the github.com/collectd org while we're at it?

**Comment by octo** [comment]
@dougm Thanks for looking into it :)

:+1: for moving *jcollectd* to the *collectd* org.

**Comment by octo** [comment]
Hi @dougm, do you have any updates on the licensing and/or moving jcollectd? Should I create a jcollectd fork under collectd/ so we can eventually transfer the repository?

**Comment by dougm** [comment]
@octo Apologies, I let this slip, just got the process started today.  I will provide updates when I have them.  Fork would be fine, though I believe we can just move the repository and github provides redirects, etc.

**Comment by octo** [comment]
@dougm Thank you so much! :-)

**Comment by dougm** [comment]
Happy to report I've received the OK to transfer ownership of jcollectd to the collectd org.
I haven't used the "Transfer ownership" feature of github before, but going to click the button now and see what happens :)

**Comment by dougm** [comment]
I thought there might be a step where the user/org can accept/decline transfer, but it just bails with "You don’t have admin rights to collectd" cc @octo

**Comment by ctubbsii** [comment]
@dougm I believe you need "owner" on both groups to transfer a repo. This may require granting temporary owner rights.

**Comment by octo** [comment]
@dougm The docs agree with @ctubbsii:

> The user transferring a repository must have owner or admin permissions in both organizations to complete a repository transfer.

I'll grant temporary permissions to you so you can do the transfer.

**Comment by dougm** [comment]
@octo ok done!

**Comment by octo** [comment]
Awesome, thanks @dougm. I've invited you to the *jcollectd* repository as an "outside collaborator".

I'll leave the issue open since the original question was about licensing.
